### PR TITLE
[Merged by Bors] - Upgraded to ICU 1.2

### PR DESCRIPTION
**Pull request by Razican**
This PR upgrades ICU to 1.2.

Unfortunately we still have some breaking changes, so this is being handled in https://github.com/unicode-org/icu4x/issues/3332

### CODE REVIEW

File: boa_engine/src/context/icu.rs
Comment (Razican) [inline boa_engine/src/context/icu.rs:10]: This change is required due to backwards in-compatibility, that might be fixed in https://github.com/unicode-org/icu4x/pull/3333, but we might just want to move to explicitly import `yoke` and `zerofrom`, so that we don't rely on `#[doc(hidden)]` features.

Comment (jedel1043) [reply boa_engine/src/context/icu.rs:10]: Yeah, it'll be better to import both deps directly.

**Comment by jedel1043** [comment]
Imported both deps directly. This is now just blocked on https://github.com/unicode-org/icu4x/pull/3335

**Comment by jedel1043** [comment]
Just realized we can completely bypass the `buffer` and `any` constructors by using the `unstable` ones.
Just for the note, `unstable` does sound ominous, but I think we're in the clear, because `BoaProvider` unconditionally implements `DataProvider<M>` for all `M: KeyedDataMarker`, which means that any added or removed `P: DataProvider<M>` bounds on the constructors should never break our compilation... I think.

@sffc Can you either confirm or deny my assumptions? I can rollback to our old method if needed, just want to be sure since doing it this way simplifies a lot our logic.

**Comment by sffc** [comment]
> @sffc Can you either confirm or deny my assumptions? I can rollback to our old method if needed, just want to be sure since doing it this way simplifies our logic a lot.

Yes, you have a blanket impl that leverages as_deserializing and as_downcasting, so you can use unstable constructors. The only catch is that the unstable constructors can't read from deprecated data keys, which we did for the first time in this release when we split the LocaleExpander key in three. So unless you regenerate your data, LocaleExpander will fail to construct. Using the non-unstable constructors allows you to use the old data key (they will map it for you internally).

```
impl<M> DataProvider<M> for BoaProvider<'_>
where
    M: KeyedDataMarker + 'static,
    for<'de> YokeTraitHack<<M::Yokeable as Yokeable<'de>>::Output>: Deserialize<'de>,
    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
    M::Yokeable: ZeroFrom<'static, M::Yokeable> + MaybeSendSync,
{
    fn load(&self, req: DataRequest<'_>) -> Result<DataResponse<M>, DataError> {
        match self {
            BoaProvider::Buffer(provider) => provider.as_deserializing().load(req),
            BoaProvider::Any(provider) => provider.as_downcasting().load(req),
        }
    }
}
```

**Comment by jedel1043** [comment]
> The only catch is that the unstable constructors can't read from deprecated data keys, which we did for the first time in this release when we split the LocaleExpander key in three.

Oh, that's pretty important, since I'd like to give our users the decision to upgrade their dataset when they want, instead of forcing them to regenerate when we bump `icu` to the latest version. I'll rollback that change.

This is still blocked on that PR then.

**Comment by Razican** [comment]
I rebased this to the latest `main` branch and squashed commits. I also tried to "patch" cargo using the branch of the PR in ICU to see if it worked, but it seems that something is not quite right, probably some double-dependency. Let's wait for the 1.2.1 release, and then we can make this proper :)

**Comment by sffc** [comment]
icu_locid_transform version 1.2.1 is released. This should unblock your upgrade

File: Cargo.toml
Comment (Razican) [inline Cargo.toml:65]:
```suggestion
codegen-units = 1

```

**Comment by Razican** [comment]
This looks good :) I would run a `cargo update` before merging (and should we re-generate the data file?), and add that end-of-file new line in `Cargo.toml`, but for the rest, I think we can merge it!

Thank you @sffc for your fast response!

**Review by jedel1043 — APPROVED** [review]
Ran datagen but git didn't detect any diff. Everything good!

**Review by HalidOdat — APPROVED** [review]
Looks good to me! :)

**Pull request closed without merging**
